doc/api/fs.md

```diff
@@ -619,6 +619,25 @@ On Linux, positional writes do not work when the file is opened in append mode.
 The kernel ignores the position argument and always appends the data to
 the end of the file.
 
+#### `filehandle.write(buffer, options)`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+* `buffer` {Buffer|TypedArray|DataView}
+* `options` {Object}
+  * `offset` {integer} **Default:** `0`
```

Comment:
Did you mean this?
```suggestion
#### `filehandle.write(options)`

<!-- YAML
added: REPLACEME
-->

* `options` {Object}
  * `buffer` {Buffer|TypedArray|DataView}
  * `offset` {integer} **Default:** `0`
```
or should we instead make `buffer` not part of the `options` object, to align with `fs.writeSync` (and add it for `fs.write`)?

Reply:
I think, each writing method must be aligned with its reading counterpart. Their current state is:
- `fs.read(fd[, options], callback)`
- `filehandle.read([options])`
- `fs.readSync(fd, buffer[, options])`

So if we want to keep same signatures for all three APIs, it might require breaking changes for reading methods.

Also, as for choosing between `{ buffer, offset, length, position }` and `buffer, { offset, length, position }`, I'm not sure if any option is comprehensively better.
When params are skippable, `{ buffer, position }` pair has a pretty atomic meaning.

As a long term (i.e. after DEP0162 EOL) solution, I'm considering adding support for both signatures simultaneously. At the cost of bloating the docs and code, it would provide a more consistent and convenient API without breaking userspace code.

Reply:
`fs.readSync` returns a `number` instead of returning `{ buffer: Buffer, bytesRead: number }`, which is why `buffer` is not optional for this method 😬
I don't think we can achieve full consistency, we have to accept a tradeoff at some point. I think it makes more sense to keep `options` for the optional parameters (so `buffer, { offset, length, position }`) because:
- it would be quite confusing to have a `options` parameter that's actually required,
- consistency between `write`* methods is more important than consistency with `read`* methods (I don't think there's a use case where someone would pass the same `options` object to both a `read` and a `write` call, is there?)

That's how I would do it:

- `fs.write(fd, buffer[, options], callback)`
- `filehandle.write(buffer[, options])`
- `fs.writeSync(fd, buffer[, options])`

Reply:
> I don't think we can achieve full consistency, we have to accept a tradeoff at some point.

What do you think about supporting both signatures in all 6 methods at the same time? Maybe even with hybrid form for better interoperability: `{ buffer, { offset, length, position } }`
It might look weird in Node core, but for userland it could be better than memorizing the differences, rewriting in case of migration to another API, etc.

A less bloated tradeoff is also possible: we can simply add `fs.read(fd, buffer[, options], callback)` and `filehandle.read(buffer[, options])`.

> - it would be quite confusing to have a `options` parameter that's actually required,

Isn't this just a terminology issue? Let's call the object with mandatory `.buffer` property `params`. Then:
`params == { buffer, ?options } == { buffer, ?{ ?offset, ?length, ?position } }`
With `fs.read`, `fh.read` "loosened" by allowing `buffer` or whole `params` to be omitted.

> - I don't think there's a use case where someone would pass the same options object to both a read and a write call, is there?

I'm afraid there should be use cases for this.
Basically whenever we have to read a part of file, modify it, and write at the same position (inplace, or in copy, or in output file, doesn't matter):
```js
let position = 0x1234;
let byteLength = 123;
let buffer = new Uint8Array(byteLength);
let options = { position /*, length: byteLength */ };
let params = { buffer, options }; // OR: { buffer, ...options }

await input_fh.read(params); // OR: .read(buffer, options) // BUT NOT: { buffer } = input.fh.read(options)
mutateBuffer(params.buffer); // OR: modifiedBuffer = await processBuffer(buffer)
await output_fh.write(params); // OR: .write(modifiedBuffer, options)
```
`options` here has a pronounced "slice on exact offset" meaning, e.g.
```js
const ipv6_destination = { position: 0x18, length: 16 };
const webp_VP8frame = { position: 0x14 };
const incrementalFileSignature = { position: fileSize - 80, length: 80 };
```
The changes do not affect string-oriented code and higher-level methods (readFile, appendFile, etc.), so I expect dealing with known internal structure of large binary files to be one of the most important uses in a scope of direct `.read()`/`.write()` invocations.